test/testdata/infer/generics/type_param_is_a.rb

```diff
@@ -0,0 +1,17 @@
+# typed: true
+
+class Test
+  extend T::Sig
+
+  sig do
+    type_parameters(:T)
+      .params(x: T.type_parameter(:T))
+      .returns(T.type_parameter(:T))
+  end
+  def test(x)
+    if x.is_a? Integer
+      T.reveal_type(x) # error: Revealed type: `T.all(Integer, Test#test#T)`
```

Comment:
Just to be safe, do you want to write method that takes an integer and make sure that within this branch it's possible to call that method, but outside it's not?

I remember us having some discussion about what the `Test#test#T` meant in this context. If you have another way to check the semantics of this rather than just the stringified form I'd be down.